stripe.go

```diff
@@ -462,32 +509,99 @@ func FormatURLPath(format string, params ...string) string {
 	return fmt.Sprintf(format, untypedParams...)
 }
 
-// GetBackend returns the currently used backend in the binding.
-func GetBackend(backend SupportedBackend) Backend {
-	switch backend {
+// GetBackend returns one of the library's supported backends based off of the
+// given argument.
+//
+// It returns an existing default backend if one's already been created.
+func GetBackend(backendType SupportedBackend) Backend {
+	var backend Backend
+
+	backends.mu.RLock()
+	switch backendType {
 	case APIBackend:
-		backends.mu.RLock()
-		ret := backends.API
-		backends.mu.RUnlock()
-		if ret != nil {
-			return ret
+		backend = backends.API
+	case UploadsBackend:
+		backend = backends.Uploads
+	}
+	backends.mu.RUnlock()
+	if backend != nil {
+		return backend
+	}
+
+	backend = GetBackendWithConfig(
+		backendType,
+		&BackendConfig{
+			HTTPClient:        httpClient,
+			LogLevel:          LogLevel,
+			Logger:            Logger,
+			MaxNetworkRetries: 0,
+			URL:               "", // Set by GetBackendWithConfiguation when empty
+		},
+	)
+
+	backends.mu.Lock()
+	defer backends.mu.Unlock()
+
+	switch backendType {
+	case APIBackend:
+		backends.API = backend
+	case UploadsBackend:
+		backends.Uploads = backend
+	}
+
+	return backend
+}
+
+// GetBackendWithConfig is the same as GetBackend except that it can be given a
+// configuration struct that will configure certain aspects of the backend
+// that's return.
+func GetBackendWithConfig(backendType SupportedBackend, config *BackendConfig) Backend {
+	if config.HTTPClient == nil {
+		config.HTTPClient = httpClient
+	}
+
+	if config.Logger == nil {
+		config.Logger = Logger
+	}
+
+	switch backendType {
+	case APIBackend:
+		if config.URL == "" {
+			config.URL = apiURL
+		}
+
+		// Add the /v1/ prefix because all client packages expect it. We should
+		// probably change this to just make it explicit wherever it's needed
+		// to fulfill the principle of least astonishment.
+		config.URL += "/v1"
+
+		return &BackendConfiguration{
```

Comment:
this like the code is identical between both switch except for the url, could we generalize that?

Reply:
Great point. I just updated the code and it looks way cleaner now.